src/LagrangianParticleTracking.jl

```diff
@@ -0,0 +1,84 @@
+module LagrangianParticleTracking
+
+export LagrangianParticles, advect_particles!
+
+using KernelAbstractions
+
+using Oceananigans.Grids
+
+using Oceananigans.Architectures: device
+using Oceananigans.Fields: interpolate, datatuple
+using Oceananigans.Utils: MAX_THREADS_PER_BLOCK
+
+import Base: size, length
+
+struct LagrangianParticles{X, Y, Z}
+    x :: X
+    y :: Y
+    z :: Z
+end
+
+function LagrangianParticles(; x, y, z)
+    size(x) == size(y) == size(z) ||
+        error("x, y, z must all have the same size!")
+        
+    (ndims(x) == 1 && ndims(y) == 1 && ndims(z) == 1) ||
+        error("x, y, z must have dimension 1 but ndims=($(ndims(x)), $(ndims(y)), $(ndims(z)))")
+
+    return LagrangianParticles(x, y, z)
+end
+
+size(particles::LagrangianParticles) = size(particles.x)
+length(particles::LagrangianParticles) = length(particles.x)
+
+@inline enforce_periodicity(x, xₗ, xᵣ, ::Type{Bounded})  = x
+
+@inline function enforce_periodicity(x, xₗ, xᵣ, ::Type{Periodic})
+    x > xᵣ && return xₗ + (x - xᵣ)
+    x < xₗ && return xᵣ - (xₗ - x)
+    return x
+end
```

Comment:
Just started with a more readable function but couldn't figure out how to rewrite using e.g. `mod` functions to get rid of the conditional statements.

Reply:
Hi Ali - currently I am using the code below for periodic domain, which is easy to broadcast. Not sure if this suits your situation.
Also, in `PlanktonIndividuals.jl` I use array broadcast instead of `KernelAbstractions.jl` for partical advection on GPU. You can see my code on `dev_gpu` branch.
```julia
x = isless(x, xᵣ) * x + isequal(x, xᵣ) * x + isless(xᵣ, x) * (x - xᵣ + xₗ)
x = isless(xₗ, x) * x + isequal(xₗ, x) * x + isless(x, xₗ) * (xᵣ - xₗ + x )
```

Reply:
Thanks @zhenwu0728 that's pretty neat! Didn't know about `isless` and `isequal`.

Reply:
Lots of work on your `dev_gpu` branch! Out of curiosity how many particles/plankton/microbes do you usually simulate? Is it a big performance hit?

Maybe I should open an issue on PlanktonIndividuals.jl to ask.

Reply:
That's OK. I usually use 1,000-10,000 particles. But I ran them on the CPU. It takes hours to run.
The GPU part is still working in progress. So I cannot tell you the benchmarks on GPU right now.  
A rough comparison was at least 3 times faster on GPU when I was testing my code.